byron/ledger/executable-spec/src/Ledger/UTxO.hs

```diff
@@ -25,82 +27,90 @@ newtype TxId = TxId { getTxId :: Hash }
 -- |The input of a UTxO.
 --
 --     * __TODO__ - is it okay to use list indices instead of implementing the Ix Type?
-data TxIn = TxIn TxId Natural deriving (Show, Eq, Ord)
+data TxIn id = TxIn id Natural deriving (Show, Eq, Ord)
 
 -- |The output of a UTxO.
-data TxOut = TxOut Addr Value deriving (Show, Eq)
+data TxOut = TxOut Addr Value deriving (Show, Eq, Ord)
 
 -- |The unspent transaction outputs.
-newtype UTxO = UTxO (Map TxIn TxOut) deriving (Show, Eq)
+newtype UTxO id = UTxO (Map (TxIn id) TxOut) deriving (Show, Eq)
 
 -- |A raw transaction
-data Tx = Tx { inputs  :: Set TxIn
-             , outputs :: [TxOut]
-             } deriving (Show, Eq)
+data Tx id = Tx { txid    :: id
+                , inputs  :: [TxIn id]
+                , outputs :: [TxOut]
+                } deriving (Show, Ord)
 
--- |Compute the id of a transaction.
-txid :: Tx -> TxId
-txid = TxId . hash
+instance Eq id => Eq (Tx id) where
+  (Tx _ ins outs) == (Tx _ ins' outs') = ins == ins' && outs == outs'
 
 -- |Compute the UTxO inputs of a transaction.
-txins :: Tx -> Set TxIn
+txins :: Tx id -> [TxIn id]
 txins = inputs
 
 -- |Compute the UTxO outputs of a transaction.
-txouts :: Tx -> UTxO
+txouts :: Ord id => Tx id -> UTxO id
 txouts tx = UTxO $ Map.fromList
   [ (TxIn transId idx, out) | (out, idx) <- zip (outputs tx) [0 ..] ]
   where transId = txid tx
 
-txfee :: Tx -> Value
-txfee = undefined
+-- TODO: Handle fees properly
+txfee :: Tx id -> Value
+txfee _ = Value 0
 
 -- |Domain restriction
 --
 --     * __TODO__ - better symbol?
-(◁) :: Set TxIn -> UTxO -> UTxO
-ins ◁ (UTxO utxo) = UTxO $ Map.filterWithKey (\k _ -> k `Set.member` ins) utxo
+(◁) :: Ord id => [TxIn id]-> UTxO id -> UTxO id
```

Comment:
I'm wondering whether we should delete these operations since we have the `Relation` class in `Ledger.Core`...

Reply:
Annoyingly we can't implement `Relation` for `UTxO` because it has the wrong kind! I can rework the type of `Relation` to use `Domain` and `Range` associated types and that might work.

Reply:
Yeah, that was what I had in mind. But I think that's outside the scope of this PR, so let's just keep this in mind.

Reply:
I went ahead and did it, as it seems to clean things up

Reply:
Damn you're fast :zap: :)